File: src/iarc7_safety/SafetyClient.cpp
////////////////////////////////////////////////////////////////////////////
//
// SafetyClient
//
// Class implements an easy way to use iarc7_safety to notify of safety events
//
////////////////////////////////////////////////////////////////////////////

// Associated header
#include "iarc7_safety/SafetyClient.hpp"

// ROS message headers

using namespace Iarc7Safety;

SafetyClient::SafetyClient(ros::NodeHandle& nh, const std::string bond_id) :
bond_id_(bond_id),
bond_("bond_topic", bond_id_, boost::bind(&SafetyClient::onBroken, this), boost::bind(&SafetyClient::onFormed, this))
{
    safety_subscriber_ = nh.subscribe("safety", 100, &SafetyClient::processSafetyMessage, this);
}

bool SafetyClient::formBond()
{
    ROS_INFO("safety_client: trying to form bond %s", bond_.getId().c_str());

    // Try to start the bond
    bond_.start();
    waitUntilSafe();

    return true;
}

// This function is a workaround since Bond::waitUntilFormed doesn't work because it doesn't spin
void SafetyClient::waitUntilSafe()
{
    while(ros::ok())
    {
        if(formed_)
        {
            break;
        }

        ros::spinOnce();

        ros::Duration(0.1).sleep();
    }
}

void SafetyClient::processSafetyMessage(const std_msgs::String::ConstPtr& message)
{
    if(message->data == bond_id_)
    {
        safety_active_ = true;
    }
    else if(message->data == fatal_message_)
    {
        safety_active_ = true;
        fatal_active_ = true;
    }
}

bool SafetyClient::isSafetyActive()
{
    return safety_active_;
}

bool SafetyClient::isFatalActive()
{
    return fatal_active_;
}

void SafetyClient::onBroken()
{
    fatal_active_ = true;
    safety_active_ = true;
    formed_ = false;
}

void SafetyClient::onFormed()
{
    formed_ = true;
}

File: src/iarc7_safety/SafetyClient.py
#!/usr/bin/env python

###########################################################################
#
# SafetyClient
#
# Class implements an easy way to use the node_monitor to notify of safety events
#
###########################################################################

import rospy
from bondpy import bondpy
from std_msgs.msg import String

class SafetyClient:
    def __init__(self, bondId):
        self.bond = bondpy.Bond("bond_topic", bondId, self.on_broken, self.on_formed)
        self.formed = False
        self.broken = False
        self.safetyActive = False
        self.fatalActive = False
        self.bondId = bondId

        self.bond.set_heartbeat_period(0.2)
        self.bond.set_heartbeat_timeout(0.5)

        rospy.Subscriber("safety", String, self.process_safety_message)

    def form_bond(self):
        self.bond.start()
        
        return self.wait_until_safe()

    def wait_until_safe(self):
        while not rospy.is_shutdown():
            if self.broken:
                return False

            if self.formed:
                return True
            rospy.sleep(0.1)

    def process_safety_message(self, message):
        
        if(message.data == self.bondId):
            self.safetyActive = True

        if(message.data == "FATAL"):
            self.fatalActive = True
            self.safetyActive = True

    def is_safety_active(self):
        return self.safetyActive

    def is_fatal_active(self):
        return self.fatalActive

    def on_broken(self):
        self.broken = True
        self.formed = False

        self.safetyActive = True
        self.fatalActive = True

    def on_formed(self):
        self.broken = False
        self.formed = True
